lib/utils/batcher/batcher.go

```diff
@@ -0,0 +1,186 @@
+//	       batcher.WithThreshold(100),
+//	   }
+//
+//		err := batcher.Run(ctx, eventChan, func(batch []Event) error {
+//			log.Printf("Processing %d events", len(batch))
+//			return processEvents(batch)
+//		}, opts...))
+//
+// The function returns when the context is canceled, returning ctx.Err(),
+// when fn returns an error, returning that error, or when the events channel is closed.
+func Run[T any](ctx context.Context, events <-chan T, fn func(batch []T) error, opts ...Option) error {
+	collector := New[T](opts...)
+	return trace.Wrap(collector.Run(ctx, events, fn))
+}
+
+// Option configures a Collector using the functional options pattern.
+type Option func(*config)
+
+// WithWindow sets the collection window duration.
+func WithWindow(d time.Duration) Option {
+	return func(c *config) {
+		if d > 0 {
+			c.window = d
+		}
+	}
+}
+
+// WithThreshold sets the maximum number of events per batch.
+// Default is unlimited (bounded only by window duration).
+func WithThreshold(n int) Option {
+	return func(c *config) {
+		if n > 0 {
+			c.threshold = n
+		}
+	}
+}
+
+// WithClock sets a custom clock for testing.
+// Default uses the system clock.
+func WithClock(clock clockwork.Clock) Option {
+	return func(c *config) {
+		if clock != nil {
+			c.clock = clock
+		}
+	}
+}
+
+type config struct {
+	window    time.Duration
+	threshold int
+	clock     clockwork.Clock
+}
+
+// New creates a new Collector with the given options.
+func New[T any](opts ...Option) *Collector[T] {
+	cfg := config{
+		window:    10 * time.Second,
+		threshold: 50,
+		clock:     clockwork.NewRealClock(),
+	}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	return &Collector[T]{config: cfg}
+}
+
+// Collector batches events into time windows.
+type Collector[T any] struct {
+	config
+}
+
+// Run starts collecting events and calls fn for each batch.
+// It blocks until ctx is canceled or an error occurs.
+func (c *Collector[T]) Run(ctx context.Context, events <-chan T, fn func(batch []T) error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		batch, err := c.CollectBatch(ctx, events)
+		if err != nil {
+			return err
+		}
+
+		if len(batch) > 0 {
+			if err := fn(batch); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+}
+
+// ErrChannelClosed is returned when the events channel is closed.
+var ErrChannelClosed = trace.BadParameter("events channel closed")
+
+// CollectBatch collects events into a batch until the time window elapses
+// or the threshold is reached, whichever comes first.
+// It blocks until at least one event is collected or ctx is canceled.
+//
+// NOTE: CollectBatch returns already collected events if ctx is canceled.
+// If you want to still process then you need to check the error returned
+// and handle ctx.Err() case separately.
+func (c *Collector[T]) CollectBatch(ctx context.Context, events <-chan T) ([]T, error) {
+	ticker := interval.New(interval.Config{
```

Comment:
`interval.Interval` is pretty heavy duty, it allocates four channels and runs a goroutine and its cost is intended to be amortized across time, it's not intended to be used literally once and then stopped.